### DISCUSSION TITLE
Fix playback to Airplay devices

### DISCUSSION THREAD

**Discussion by marcelveldt** [music-assistant/support#438] [Feature requests and ideas]
The current state of the atv integration in HA is that it won't stream audio to Airplay speakers.
An official ATV 4K does work (sometimes) and Homepod is hit and miss.

Maybe we can fix something to send the stream directly to the pyatv library because MA is using a raw PCM stream internally.

**Comment by Mariusthvdb** [comment]
just letting you know that on v1.0.17, and trying to play a Tunein Radio 4 Nl stream to my HomePod (select the stream in Radio, click Play in the bottom control section), I get:
```
Logger: homeassistant.core
Source: components/apple_tv/media_player.py:297 
First occurred: 08:18:58 (1 occurrences) 
Last logged: 08:18:58

Error executing service: <ServiceCall media_player.play_media (c:01G2VEQ90PPFC8QZKSV90KZJVF): entity_id=['media_player.homepod_keuken'], extra=, media_type=music, media_id=http://192.168.1.71:8095/media_player.homepod_keuken.flac>
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/core.py", line 1662, in catch_exceptions
    await coro_or_task
  File "/usr/src/homeassistant/homeassistant/core.py", line 1681, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 204, in handle_service
    await service.entity_service_call(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 690, in entity_service_call
    future.result()  # pop exception if have
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 964, in async_request_call
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 728, in _handle_entity_call
    await result
  File "/usr/src/homeassistant/homeassistant/components/apple_tv/media_player.py", line 297, in async_play_media
    await self.atv.stream.stream_file(media_id)
  File "/usr/local/lib/python3.9/site-packages/pyatv/core/facade.py", line 382, in stream_file
    await self.relay("stream_file")(file, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/pyatv/protocols/raop/__init__.py", line 341, in stream_file
    self.playback_manager.acquire()
  File "/usr/local/lib/python3.9/site-packages/pyatv/protocols/raop/__init__.py", line 141, in acquire
    raise exceptions.InvalidStateError("already streaming to device")
pyatv.exceptions.InvalidStateError: already streaming to device

```

Even while nothing was playing at all.... not the stream seems to be `media_player.homepod_keuken.flac` ?

**Comment by marcelveldt** [comment]
ATV streaming implementation is flaky. I think the only way would be if MA is going to send the stream to airplay devices itself.

**Comment by Mariusthvdb** [comment]
for MA I guess that would safest. For HA, the ATV is a horror, and should be fixed nonetheless... the amount of errors is endless

**Comment by marcelveldt** [comment]
Agreed

**Comment by erkr** [comment]
Moved to discussions, raise your votes!

**Comment by ddppddpp** [comment]
I think issues are not limited to MA or to the ATV implementation in HA. Interacting with AirPlay based media players is not really there. By AirPlay based I mean both Apple manufactured (atv, homepods), apple-licensed (i.e. Marantz/Denon receivers, tv's etc.) and OSS-based (i.e. shairport-sync).

**Reply by marcelveldt** [reply]
No, streaming to Airplay capable devices is perfectly possible just not with the current HA implementation which is more focused on the Apple TV side of things instead of actual music streaming.

Maybe I'll just create an additional airplay component...

**Comment by postlund** [comment]
I can add a comment here. (as I'm the author of pyatv and the corresponding Apple TV integration). The reason Home Assistant cannot stream to plain AirPlay devices is because pyatv only supports AirPlay v1 devices. Since a lot (especially new) devices only support AirPlay v2, we decided to add a filter to not discover devices that aren't Apple devices. Ideally we would only ignore pure AirPlay v2 devices, but it's harder to do then one might think... My intention is to support AirPlay v2 devices too at some point, but I'm not sure when that will happen. I have basics ready, like setting up streams, but something still isn't working with the encryption of streamed audio.

The filter only applies to discovery though, so it should be possible to add a device manually via IP (assuming the device support AirPlay v1). The streaming support is a bit unstable as it's very early in the development process. Especially when streaming from a remote source, e.g. via https (which is used by TTS) it will likely fail in a lot of cases. There's a very little buffer and pyatv does not handle buffer underruns very good.

**Reply by marcelveldt** [reply]
@postlund great to see you reply here!
It's not only Airplay 2 devices that are not working. Streaming audio is not possible at all to any Airplay or ATV device.
I've briefly looked at the code and spotted what you describe. It can however be less painful to skip the transcoding in pyatv and just let MA send the raw PCM (or ALAC encoded) stream to the airplay devices ? That's the direction I was thinking about. Maybe even wrap pyatv in a new integration called airplay to make it more clear that is focused on audio ?

Anyways, I'd love to work with you to make airplay work. I believe it would be a very good addition

**Reply by postlund** [reply]
Can you elaborate in more detail what isn't working? I might be able to give you better answers then, as AirPlay receivers *should* work, assuming they support AirPlay v1 as mentioned. In Home Assistant it's only possible to pass a file or URL though. When using pyatv natively (via API), you can pass something FILE-like and that should work (with transcoding if needed). I will definitely help out as much as I can to make it happen!

I **strongly** recommend not creating a new integration based on pyatv. Getting the everything right in regards to scanning, pairing and credentials handling is not for the faint of heart. I've gotten it pretty decent (and generic) now, so it would be better to improve upon what we already have.

**Reply by marcelveldt** [reply]
Great to hear! I'm pretty sure that we can make this work with combined forces. 
It's been a while since I last tried but I will for sure try it all out again later this week and note the issues. This is what I remember:

- Playback of URL only works one time on ATV, regardless if its a radio stream, TTS or Music Assistant.
- Airplay speaker (tried a few I have here) - added manually by IP worked fine. Playback did not work at all with an error that playback is not supported on the device.

Music Assistant works internally with raw PCM audio and then transcodes the audio to whatever codec the device prefers, default is FLAC. Audio delivery happens by default over HTTP so the player just receives a local/internal path to the stream for that player.
We can also make a different route for the atv integration and hook it into the audio callback directly so it receives the pcm chunks directly to bypass the decoder in pyatv. It makes no sense to have PCM audio, encode to FLAC, send it to pyatv which in turn decodes it again and encodes it. Better to just skip that step (if even possible).

**Reply by G-Christ** [reply]
Hi, 
None of my Airplay 1 devices are displayed on the Music Assistant setting page where I can select media players.

This was though possible in some earlier version of the Music Assistant.

How can I add these manually or by any means?

Thanks.

**Comment by postlund** [comment]
Regarding the first point, you might have been using the `play_url` API call in that case. It doesn't use streaming (via RAOP) but rather the `/play` endpoint. There's a bug in tvOS that breaks everything (metadata and all sorts of things) after the first playback. A reboot of the Apple TV is required. Apple does not seem to want to fix it.

The second bullet is likely because of other bugs in the `play_media` logic, hopefully it works now.

**Comment by ohhai-kthxbai** [comment]
Hello - Is airplay (v1) support currently (as of 2022.8.4) broken, or a new feature?
I see no open issue but don't want to start one if it's not a feature!

**Reply by OzGav** [reply]
No it is not a feature. This thread is the discussion to suggest it’s implementation

**Comment by Schlichi83** [comment]
Is there any progress on this? I have lots of HomePods in my house and I would love to be able to stream from Home Assistant / Music Assistant to them.

**Comment by kimme1024** [comment]
Same here.
The 2 devices I use most are a HomePod mini and a Yamaha receiver with AirPlay. I hope AirPlay support will be added so I can start using Music Assistant.

**Reply by jackoncept** [reply]
Would love it too !!!

**Comment by marcelveldt** [comment]
MA 2.0 beta has Airplay support. this feature request can be considered done.
